zebra-network/src/peer_set/initialize/recent_by_ip.rs

```diff
@@ -74,13 +79,12 @@ impl RecentByIp {
         // which is unexpected, but stops this list growing without limit.
         // After the handshake, the peer set will remove any duplicate connections over the limit.
         let age_limit = now - self.time_limit;
+        // This is the lowest IP address among all IpAddr.
+        let dummy_ip = Ipv4Addr::UNSPECIFIED.into();
 
-        // `by_time` must be sorted for this to work.
-        let split_off_idx = self.by_time.partition_point(|&(_, time)| time <= age_limit);
-
-        let updated_by_time = self.by_time.split_off(split_off_idx);
+        let updated_by_time = self.by_time.split_off(&(age_limit, dummy_ip));
```

Comment:
This seems to check every element:

![image](https://github.com/ZcashFoundation/zebra/assets/5491350/28081acd-61a6-4509-9196-d9f7337f601d)

Reply:
You're right, that's weird:
> Currently, our implementation simply performs naive linear search. 

https://doc.rust-lang.org/std/collections/struct.BTreeMap.html

However, due to caching, performance is usually ok:
> This provides excellent performance on small nodes of elements which are cheap to compare. However in the future we would like to further explore choosing the optimal search strategy based on the choice of B, and possibly other factors. Using linear search, searching for a random element is expected to take B * log(n) comparisons, which is generally worse than a BST. In practice, however, performance is excellent.

But this is unlikely to hold for large numbers of elements.

And a `VecDeque` deals with the cache-miss part of the issues with binary trees:
> In theory, a binary search tree (BST) is the optimal choice for a sorted map, as a perfectly balanced BST performs the theoretical minimum amount of comparisons necessary to find an element (log2n). However, in practice the way this is done is very inefficient for modern computer architectures. In particular, every element is stored in its own individually heap-allocated node. This means that every single insertion triggers a heap-allocation, and every single comparison should be a cache-miss. Since these are both notably expensive things to do in practice, we are forced to, at the very least, reconsider the BST strategy.

Reply:
So I don't think we should merge this, thanks for checking!